service/collector/nat.go

```diff
@@ -0,0 +1,270 @@
+	awsClientsList, err := v.helper.GetAWSClients(reconciledClusters)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for _, item := range awsClientsList {
+		awsClients := item
+
+		err := v.collectForAccount(ch, awsClients)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	return nil
+}
+
+func (v *NAT) Describe(ch chan<- *prometheus.Desc) error {
+	ch <- natDesc
+	return nil
+}
+
+func (v *NAT) collectForAccount(ch chan<- prometheus.Metric, awsClients clientaws.Clients) error {
+	accountID, err := v.helper.AWSAccountID(awsClients)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	var natInfo *natInfoResponse
+	// Check if response is cached
+	natInfo, err = v.cache.Get(accountID)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	//Cache empty, getting from API
+	if natInfo == nil || natInfo.vpcs == nil {
+		natInfo, err = getNatInfoFromAPI(accountID, awsClients)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	if natInfo != nil {
+		err = v.cache.Set(accountID, *natInfo)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		for vpcID, vpcInfo := range natInfo.vpcs {
+			for azName, azValue := range vpcInfo.natGatewaysByZone {
+				ch <- prometheus.MustNewConstMetric(
+					natDesc,
+					prometheus.GaugeValue,
+					azValue,
+					accountID,
+					vpcID,
+					azName,
+				)
+			}
+		}
+	}
+
+	return nil
+}
+
+// getNatInfoFromAPI collect from AWS API the number of NAT Gateways by Availability Zone for
+// each VPC of the installation
+func getNatInfoFromAPI(accountID string, awsClients clientaws.Clients) (*natInfoResponse, error) {
+	var res natInfoResponse
+	res.vpcs = make(map[string]vpcInfo)
+
+	// 1. Get all VPCs of the installation
+	iv := &ec2.DescribeVpcsInput{
+		Filters: []*ec2.Filter{
+			{
+				Name: aws.String(fmt.Sprintf("tag:%s", key.TagOrganization)),
+				Values: []*string{
+					aws.String("giantswarm"),
+				},
+			},
+		},
+	}
+	rv, err := awsClients.EC2.DescribeVpcs(iv)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	// 2. Get all NAT GWs for each VPC
+	for _, vpc := range rv.Vpcs {
+		res.vpcs[*vpc.VpcId] = vpcInfo{
+			natGatewaysByZone: make(map[string]float64),
+		}
+
+		in := &ec2.DescribeNatGatewaysInput{
+			Filter: []*ec2.Filter{
+				{
+					Name: aws.String("vpc-id"),
+					Values: []*string{
+						vpc.VpcId,
+					},
+				},
+			},
+		}
+		rn, err := awsClients.EC2.DescribeNatGateways(in)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		// 3. Get all the subnets for each NAT GW
+		for _, nat := range rn.NatGateways {
+			is := &ec2.DescribeSubnetsInput{
+				SubnetIds: []*string{
+					nat.SubnetId,
+				},
+			}
+			rs, err := awsClients.EC2.DescribeSubnets(is)
+			if err != nil {
+				return nil, microerror.Mask(err)
+			}
+
+			// 4. Store the number of GWs by Availability Zone
+			for _, sub := range rs.Subnets {
+				zoneID := *sub.AvailabilityZoneId
+				if _, ok := res.vpcs[*vpc.VpcId].natGatewaysByZone[zoneID]; ok {
```

Comment:
```suggestion
				if _, exists := res.vpcs[*vpc.VpcId].natGatewaysByZone[zoneID]; exists {
```
Maybe slightly easier to read through?